### PR TITLE
Support persisting docs for `materialized_view` materializations

### LINKED ISSUE #11: [Bug] Materialized View Documentation fails in PostgreSQL
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When deploying to PostgreSQL, running a model for a Materialized View that has documentation in a schema.yml associated, it tries to run the command "comment on materialized_view ...." and PostgreSQL expects "comment on materialized view ... " (without the underscore character) otherwise it throws Syntax Error - which is happening now.

The problem is that DBT model config type for MV materialization is defined as "materialized_view" and this is not translated as "comment on materialized view ... " when deploying the comment on PostgreSQL.

image
image
image
image
image
image

### Expected Behavior

DBT generates the correct comment command for PostgreSQL for Materialzied Views, which is "comment on materialized view ..." (without the underscore)

### Steps To Reproduce

1- Create a model with materialization of materialized_view
2- Add description with a schema.yml
3- Execute dbt run -s <model> configured to deploy on a PostgreSQL server

### Relevant log output

```shell
0m17:20:21.111539 [error] [MainThread]:   Database Error in model mv_test (models/mv_test.sql)
  syntax error at or near "materialized_view"
  LINE 5:   comment on materialized_view "my_database"."...
                       ^
  compiled Code at target/run/dbt_proj/models/mv_test.sql
[0m17:20:21.111773 [info ] [MainThread]: 
[0m17:20:21.112015 [info ] [MainThread]: Done. PASS=0 WARN=0 ERROR=1 SKIP=0 TOTAL=1
```

### Environment

```markdown
- OS: macOS Sonoma v14.3
- Python: 3.10.6
- dbt: 1.7.7
```

### Which database adapter are you using with dbt?

postgres

### Additional Context

_No response_

### LINKED ISSUE #120: [Bug] `persist_docs` does not work for `materialized_view` materializations
### Is this a new bug?

- [X] I believe this is a new bug
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

The macro that backs object-level `persist_docs` uses `{{ relation.type }}` to derive the object type in the `COMMENT ON` statement. This breaks for `materialized_view` materializations:

```bash
15:35:01  Database Error in model funds_movement (models/marts/funds_movement.sql)
15:35:01    Expected one of TABLE or VIEW or COLUMN or MATERIALIZED or SOURCE or SINK or INDEX or FUNCTION or CONNECTION or TYPE or SECRET or ROLE or DATABASE or SCHEMA or CLUSTER, found identifier "materialized_view"
15:35:01    LINE 5:   comment on materialized_view "marta"."public"."funds_moveme...
15:35:01                         ^
15:35:01    compiled Code at target/run/mz_get_started/models/marts/funds_movement.sql
15:35:01
15:35:01  Done. PASS=3 WARN=0 ERROR=2 SKIP=0 TOTAL=5
```

### Expected Behavior

The `persist_docs` config should work for `materialized_view` materializations.

### Steps To Reproduce

Attempt to use the `persist_docs` config with a materialized view model.

### Relevant log output

_No response_

### Environment

_No response_

### Additional Context

We've fixed this a while back for `dbt-materialize` (which shims `dbt-postgres`) in [Materialize #21878](https://github.com/MaterializeInc/materialize/pull/21878), but it really should be fixed at the source! This issue was also reported by a user of the `dbt-risingwave` adapter ([`dbt-risingwave` #31](https://github.com/risingwavelabs/dbt-risingwave/issues/31)), which also shims `dbt-postgres`.

### PR DESCRIPTION
**Pull request by morsapaes**
The macro that backs object-level `persist_docs` uses `{{ relation.type }}` to derive the object type in the `COMMENT ON` statement. This breaks for `materialized_view` materializations:

```bash
15:35:01  Database Error in model funds_movement (models/marts/funds_movement.sql)
15:35:01    Expected one of TABLE or VIEW or COLUMN or MATERIALIZED or SOURCE or SINK or INDEX or FUNCTION or CONNECTION or TYPE or SECRET or ROLE or DATABASE or SCHEMA or CLUSTER, found identifier "materialized_view"
15:35:01    LINE 5:   comment on materialized_view "marta"."public"."funds_moveme...
15:35:01                         ^
15:35:01    compiled Code at target/run/mz_get_started/models/marts/funds_movement.sql
15:35:01
15:35:01  Done. PASS=3 WARN=0 ERROR=2 SKIP=0 TOTAL=5
```

We've fixed this a while back for `dbt-materialize` (which shims `dbt-postgres`) in [Materialize #21878](https://github.com/MaterializeInc/materialize/pull/21878), but it really should be fixed at the source!

<hr>

Fixes #120, fixes #11.

### CODE REVIEW

**Comment by morsapaes** [comment]
Signed the CLA, but...not sure if some manual steps are needed to thread that through?

**Review by dataders — COMMENTED** [review]
@morsapaes thanks! just did some stuff and the tests are running now.

I see that dbt-materialize inherits from dbt-postgres. Does this proposal fix both a postgres issue and materialize issue?

Last -- can you open (or link to) an issue on this repo? This is just for tracking purposes. cheers!

**Comment by morsapaes** [comment]
> Does this proposal fix both a postgres issue and materialize issue?

It fixes a `dbt-postgres` issue that is inherited by our adapter (and others shimming `dbt-postgres`). We put up a patch a while back in `dbt-materialize` ([Materialize #21878](https://github.com/MaterializeInc/materialize/pull/21878)), since it was blocking our users.

> Two things. Can you:
> make a changelog entry with changie ([docs](https://github.com/dbt-labs/dbt-postgres/blob/main/CONTRIBUTING.md#changelog-entry))? Integration tests are blocked for that reason
open (or link to) an issue on this repo? This is just for tracking purposes?

Opened #120 for tracking purposes, and marked this PR as fixing it. Will push a changelog entry shortly, too.

Thanks, @dataders!

File: dbt/include/postgres/macros/adapters.sql [hunk 1]
```diff
@@ -196,7 +196,12 @@
 
 {% macro postgres__alter_relation_comment(relation, comment) %}
   {% set escaped_comment = postgres_escape_comment(comment) %}
-  comment on {{ relation.type }} {{ relation }} is {{ escaped_comment }};
+  {% if relation.type = 'materialized_view' -%}
+    {% set relation_type = "materialized view" %}
+  {%- else -%}
+    {%- set relation_type = relation.type -%}
+  {%- endif -%}
+  comment on {{ relation_type }} {{ relation }} is {{ escaped_comment }};
```
Comment (dataders) [inline dbt/include/postgres/macros/adapters.sql:204]: [tests are currently failing with below error](https://github.com/dbt-labs/dbt-postgres/actions/runs/9681773092/job/26719362444?pr=98#step:7:73417)

```py
ERROR tests/functional/unit_testing/test_ut_sources.py::TestUnitTestSourceInput::test_source_input - dbt_common.exceptions.base.CompilationError: Compilation Error
  expected token 'end of statement block', got '='
    line 3
      {% if relation.type = 'materialized_view' -%}
```

Comment (morsapaes) [reply dbt/include/postgres/macros/adapters.sql]: This is now solved, all tests are passing.

**Comment by mikealfare** [comment]
Thanks for the contribution @morsapaes! Do you mind adding a test case for this? We want to at least make sure that a user can submit a materialized view model with a description. Bonus points for verifying that the materialized view has the expected comment.

**Comment by morsapaes** [comment]
I will as soon as I have some bandwidth this week, @mikealfare.

**Comment by colin-rogers-dbt** [comment]
@morsapaes took a quick pass at some tests, we can open an issue to expand out base adapter persist docs tests to cover materialized views

**Review by colin-rogers-dbt — APPROVED** [review]

**Pull request merged**